### PR TITLE
Fix read batching message by pulsar reader

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

Fix bug of reader.hasMessageAvailable().

```java
while (reader.hasMessageAvailable()) {
        Message msg = reader.readNext();
    // Do something
}
```

If lastDequeuedMessage with a batchMessageId, reader.hasMessageAvailable() will return false after read first message of batchMessage, because compared message id is a MessageIdImpl(lastMessageIdInBroker) 

### Modifications

Add batching message check.

### Verifying this change

- [x] This change added tests and can be verified as follows:
- *Added UT tests for read batch message and non-batch message*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by sijie** [comment]
run java8 tests
run integration tests

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 1]
```diff
@@ -1431,6 +1428,15 @@ public boolean hasMessageAvailable() throws PulsarClientException {
         return booleanFuture;
     }
 
+    private boolean hasMoreMessages(MessageId lastMessageIdInBroker, MessageId lastDequeuedMessage) {
+        return (lastMessageIdInBroker.compareTo(lastDequeuedMessage) > 0 ||
+                // Make sure batch message can be read completely.
+                lastDequeuedMessage instanceof BatchMessageIdImpl
+                        && lastMessageIdInBroker.compareTo(lastDequeuedMessage) >= 0
+                        && incomingMessages.size() > 0)
+                && ((MessageIdImpl)lastMessageIdInBroker).getEntryId() != -1;
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java]: I think the condition is not correct in current form. I believe it should be like : 

```java
if (lastMessageIdInBroker.compareTo(lastDequeuedMessage) > 0 
    ((MessageIdImpl)lastMessageIdInBroker).getEntryId() != -1) {
    return true;
} else if (lastMessageIdInBroker.compareTo(lastDequeuedMessage) == 0
               && incomingMessages.size() > 0 ) {
     return true;
} else {
   return false;
}
```

Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java]: @merlimat Thanks, more clearer.

**Comment by merlimat** [comment]
run java8 tests
run integration tests

**Comment by sijie** [comment]
run java8 tests
run integration tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by merlimat** [comment]
run integration tests

**Comment by merlimat** [comment]
Merged in 2.3.1 at 
8e3f014

**Pull request merged**